File: internal/installer/installed.go
package installer

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/gvcgo/goutils/pkgs/gutils"
	"github.com/gvcgo/version-manager/internal/luapi/lua_global"
	"github.com/gvcgo/version-manager/internal/luapi/plugin"
)

/*
1. Find current version.
2. Find installed versions.
*/

type InstalledVersionFinder struct {
	PluginName        string
	SDKName           string
	InstalledVersions []string
	CurrentVersion    string
	Installer         *Installer
}

func NewIVFinder(sdkName string) (i *InstalledVersionFinder) {
	pls := plugin.NewPlugins()
	pls.LoadAll()
	p := pls.GetPluginBySDKName(sdkName)

	versions := plugin.NewVersions(p.PluginName)
	if versions == nil {
		return nil
	}

	_, vItem := versions.GetLatestVersion()

	i = &InstalledVersionFinder{
		PluginName: p.PluginName,
		SDKName:    sdkName,
		Installer:  NewInstaller(sdkName, p.PluginName, "", lua_global.Item{Installer: vItem.Installer}),
	}
	return
}

func (i *InstalledVersionFinder) findCurrentVersion(symbolPath string) {
	if ok, _ := gutils.PathIsExist(symbolPath); !ok {
		return
	}

	slink, _ := os.Readlink(symbolPath)
	fName := filepath.Base(slink)
	namePrefix := fmt.Sprintf("%s-", i.PluginName)
	if strings.HasPrefix(fName, namePrefix) {
		i.CurrentVersion = strings.TrimPrefix(fName, namePrefix)
	}
}

func (i *InstalledVersionFinder) FindAll() (r []string, current string) {
	current = ""
	sdkInstaller := i.Installer.GetSDKInstaller()
	if sdkInstaller == nil {
		return
	}
	symbolPath := sdkInstaller.GetSymbolLinkPath()
	versionDir := filepath.Dir(symbolPath)
	if ok, _ := gutils.PathIsExist(versionDir); !ok {
		return
	}
	i.findCurrentVersion(symbolPath)

	namePrefix := fmt.Sprintf("%s-", i.PluginName)
	dList, _ := os.ReadDir(versionDir)
	for _, d := range dList {
		if d.IsDir() && strings.HasPrefix(d.Name(), namePrefix) {
			i.InstalledVersions = append(i.InstalledVersions, strings.TrimPrefix(d.Name(), namePrefix))
		}
	}
	return i.InstalledVersions, i.CurrentVersion
}

func (i *InstalledVersionFinder) UninstallAllVersions() {
	sdkInstaller := i.Installer.GetSDKInstaller()
	if sdkInstaller == nil {
		return
	}
	versionDir := filepath.Dir(sdkInstaller.GetSymbolLinkPath())
	if ok, _ := gutils.PathIsExist(versionDir); !ok {
		return
	}
	os.RemoveAll(versionDir)
	i.Installer.UnsetEnv()
}

File: rust-vmr/crates/vmr-install/src/installed.rs
use std::fs;
use crate::common;

/// Finds installed SDK versions and the current (symlinked) version.
pub struct InstalledVersionFinder {
    pub plugin_name: String,
    pub installed_versions: Vec<String>,
    pub current_version: String,
}

impl InstalledVersionFinder {
    pub fn new(plugin_name: &str) -> Self {
        InstalledVersionFinder {
            plugin_name: plugin_name.to_string(),
            installed_versions: Vec::new(),
            current_version: String::new(),
        }
    }

    /// Find current version from the symlink target.
    fn find_current_version(&mut self, symbol_path: &std::path::Path) {
        if !symbol_path.exists() {
            return;
        }
        // Read symlink target
        if let Ok(target) = fs::read_link(symbol_path) {
            if let Some(fname) = target.file_name().and_then(|n| n.to_str()) {
                let prefix = format!("{}-", self.plugin_name);
                if let Some(ver) = fname.strip_prefix(&prefix) {
                    self.current_version = ver.to_string();
                }
            }
        }
    }

    /// Find all installed versions and current version.
    /// Returns (installed_versions, current_version).
    pub fn find_all(&mut self, sdk_name: &str) -> (&[String], &str) {
        let version_dir = common::get_sdk_version_dir(sdk_name);
        if !version_dir.exists() {
            return (&self.installed_versions, &self.current_version);
        }

        // Find current from symlink
        let symbol_path = version_dir.join(sdk_name);
        self.find_current_version(&symbol_path);

        let prefix = format!("{}-", self.plugin_name);
        if let Ok(entries) = fs::read_dir(&version_dir) {
            for entry in entries.flatten() {
                if let Ok(ft) = entry.file_type() {
                    if ft.is_dir() {
                        let name = entry.file_name();
                        let name_str = name.to_string_lossy();
                        if let Some(ver) = name_str.strip_prefix(&prefix) {
                            self.installed_versions.push(ver.to_string());
                        }
                    }
                }
            }
        }

        (&self.installed_versions, &self.current_version)
    }

    /// Uninstall all versions for this plugin.
    pub fn uninstall_all_versions(&self, sdk_name: &str) {
        let version_dir = common::get_sdk_version_dir(sdk_name);
        if version_dir.exists() {
            let _ = fs::remove_dir_all(&version_dir);
        }
    }
}
